stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+#define STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <tuple>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+constexpr int static_sum() { return 0; }
+
+template <typename Arg0, typename... Args>
+constexpr int static_sum(Arg0 arg0, Args... args) {
+  return arg0 + static_sum(args...);
+}
+
+}  // namespace internal
+
+/**
+ * Represents multiple exprs that will be calculated in same loop.
+ * @tparam T_expressions types of exprs
+ */
+template <typename... T_expressions>
+class eigen_expressions_ {
+ public:
+  /**
+   * Constructor.
+   * @param exprs expresions that will be calculated in same loop.
+   * @throw invalid_argument expressions have different sizes (for row major
+   * expressions rows and columns are swapped for the check)
+   */
+  explicit eigen_expressions_(T_expressions&&... exprs)
+      : exprs_(std::forward<T_expressions>(exprs)...) {
+    index_apply<sizeof...(T_expressions) - 1>([&](auto... Is) {
```

Comment:
[optional] should we have something like a static assert here checking these are all Eigen types? That would just let it fail with a nice error message if any non-eigen types are brought in

Reply:
There are requires on helper functions that are used for constructing `expressions_` and `results_` objects. In practice constructions will always go through these helpers.